src/libYARP_OS/src/NestedContact.cpp

```diff
@@ -53,33 +62,137 @@ bool NestedContact::fromString(const ConstString& nFullName) {
             }
             if (ok) {
                 category += ch;
-                nestedName = nestedName.substr(0, nestedName.length()-offset);
+                nestedName = nestedName.substr(0, nestedName.length() - offset);
             }
         }
         return true;
     }
     idx = fullName.find('=');
-    if (idx!=ConstString::npos) {
+    if (idx != ConstString::npos) {
         nodeName = fullName.substr(0, idx);
-        nestedName = fullName.substr(idx+1, fullName.length());
+        nestedName = fullName.substr(idx + 1, fullName.length());
         idx = nestedName.find('/');
         if (idx != ConstString::npos) {
-            if (idx==0) return true;
+            if (idx == 0)
+                return true;
             category = nestedName.substr(0, idx);
             nestedName = nestedName.substr(idx, nestedName.length());
             return true;
         }
     }
     idx = fullName.find('#');
-    if (idx!=ConstString::npos) {
+    if (idx != ConstString::npos) {
         nodeName = fullName.substr(0, idx);
-        nestedName = fullName.substr(idx+1, fullName.length());
-        char ch = nodeName[nodeName.length()-1];
-        if (ch=='-'||ch=='+') {
+        nestedName = fullName.substr(idx + 1, fullName.length());
+        char ch = nodeName[nodeName.length() - 1];
+        if (ch == '-' || ch == '+') {
             category += ch;
-            nodeName = nodeName.substr(0, nodeName.length()-1);
+            nodeName = nodeName.substr(0, nodeName.length() - 1);
         }
         return true;
     }
     return false;
 }
+
+
+NestedContact::NestedContact() :
+        mPriv(new Private())
+{
+}
+
+NestedContact::NestedContact(const ConstString& fullName) :
+        mPriv(new Private())
+{
+    fromString(fullName);
+}
+
+NestedContact::NestedContact(const NestedContact& rhs) :
+        mPriv(new Private(*(rhs.mPriv)))
+{
+}
+
+NestedContact::NestedContact(NestedContact&& rhs) :
+        mPriv(new Private(std::move(*(rhs.mPriv))))
```

Comment:
I am a bit confused by this.  The move constructor should have the advantage to avoid the `new`?

Reply:
good catch...
Probably it should just be

```c++
NestedContact::NestedContact(NestedContact&& rhs) :
        mPriv(std::move(rhs.mPriv))
```

Reply:
Probably also `NestedContact::NestedContact(NestedContact&& rhs) = default` should work...

Reply:
I don't think `mPriv(std::move(rhs.mPriv))` will work, see https://stackoverflow.com/questions/9452503/why-does-moving-a-pointer-variable-not-set-it-to-null . I guess the default version should work, but I am not really sure about how that will work.

Reply:
@traversaro now I remember the problem...
The problem was that the values are swapped, therefore the rvalue pointer is replaced with an uninitialized pointer, then it is deleted, hence crash.
I replaced it with the `new Private`, but of course this does not make sense for the move constructor.
I think this should work:
```c++
NestedContact::NestedContact(NestedContact&& rhs) :
        mPriv(std::move(rhs.mPriv))
{
    rhs.mPriv = nullptr;
}
```
@traversaro does it look ok to you?

Reply:
BTW, at some point I'd like to come to this... https://oliora.github.io/2015/12/29/pimpl-and-rule-of-zero.html